### ISSUE TITLE
Quarkus-style configuration of pact location and broker

### ISSUE THREAD

**Issue by holly-cummins** [quarkiverse/quarkus-pact#9]
We should make sure things set in application.properties are made available to the tests. 

This is related to https://github.com/quarkiverse/quarkus-pact/issues/3, but different enough that it should get its own issue. The other issue is about enabling dynamic behaviour in the annotation, this is about more general quarkus-friendly config, as we do for many other extensions.

**Comment by holly-cummins** [comment]
This should be quite straightforward and perhaps even a no-op (apart from testing), since the broker config can be set by setting the `pactbroker.*` system properties

```
  /**
   * @return URL of pact broker
   */
  String url() default "${pactbroker.url:}";
 ```

**Comment by holly-cummins** [comment]
I retract my earlier optimism. The Pact classes read directly from the system properties, which means they wouldn't see anything set in `application.properties`. We may need changes to Pact itself to allow injection of config, or we would need to instrument the annotations to inject our config. 

We'd also need to think about precedence if config is set with a system property and in application properties (which Quarkus config may automatically sort out for us), and if it's set in an annotation and in application properties (which Quarkus config would not help with).

**Comment by kmscheuer** [comment]
@holly-cummins I used a custom ValueResolver for using the application.properties.file:

`package com.example;

import au.com.dius.pact.core.support.expressions.ValueResolver;
import org.eclipse.microprofile.config.ConfigProvider;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;
import java.util.Optional;

public class QuarkusPactResolver implements ValueResolver
{
	@Override
	public boolean propertyDefined(@NotNull String s)
	{
		return getOptionalValue(s).isPresent();
	}

	private static Optional<String> getOptionalValue(String key)
	{
                	String cleanedKey = Objects.toString(key, ":").split(":")[0];
		return ConfigProvider.getConfig().getOptionalValue(cleanedKey, String.class);
	}

	@Override
	public @Nullable String resolveValue(@Nullable String s)
	{
		return getOptionalValue(s).orElse(null);
	}

	@Override
	public @Nullable String resolveValue(@Nullable String s, @Nullable String s1)
	{
		return getOptionalValue(s).orElse(s1);
	}
}
`

or for the .env file:

`package com.example;

import au.com.dius.pact.core.support.expressions.ValueResolver;
import io.github.cdimascio.dotenv.Dotenv;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;
import java.util.Optional;

public class DotEnvPactResolver implements ValueResolver
{
	@Override
	public boolean propertyDefined(@NotNull String s)
	{
		return getOptionalValue(s).isPresent();
	}

	private static Optional<String> getOptionalValue(String key)
	{
		String cleanedKey = Objects.toString(key, ":").split(":")[0];
		return Optional.ofNullable(Dotenv.load().get(cleanedKey));
	}

	@Override
	public @Nullable String resolveValue(@Nullable String s)
	{
		return getOptionalValue(s).orElse(null);
	}

	@Override
	public @Nullable String resolveValue(@Nullable String s, @Nullable String s1)
	{
		return getOptionalValue(s).orElse(s1);
	}
}`

Usage:
@PactBroker(url = "https://pact-broker.example.com", valueResolver = DotEnvPactResolver.class)

**Comment by holly-cummins** [comment]
@kmscheuer Ohh, that's a good idea! I/we could probably insert a Quarkus-aware resolver automatically once https://github.com/orgs/quarkusio/projects/30 is done.

**Comment by kmscheuer** [comment]
Refined and combined it a bit, maybe that's helpful:
```
package com.example;

import au.com.dius.pact.core.support.expressions.ValueResolver;
import io.github.cdimascio.dotenv.Dotenv;
import org.eclipse.microprofile.config.ConfigProvider;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;
import java.util.Optional;

public class DotEnvQuarkusPactResolver implements ValueResolver
{
	@Override
	public boolean propertyDefined(@NotNull String s)
	{
		return getOptionalValue(s).isPresent();
	}

	private static Optional<String> getOptionalValue(String key)
	{
		String cleanedKey = Objects.toString(key, ":").split(":")[0];

		return getDotEnvValue(cleanedKey)
			.or(() -> getQuarkusProperty(cleanedKey))
			.or(() -> getSystemProperty(cleanedKey));
	}

	private static @NotNull Optional<String> getSystemProperty(String cleanedKey)
	{
		return Optional.ofNullable(System.getProperty(cleanedKey));
	}

	private static Optional<String> getQuarkusProperty(String cleanedKey)
	{
		return ConfigProvider.getConfig().getOptionalValue(cleanedKey, String.class);
	}

	private static @NotNull Optional<String> getDotEnvValue(String cleanedKey)
	{
		return Optional.ofNullable(Dotenv.configure().ignoreIfMissing().load().get(cleanedKey));
	}

	@Override
	public @Nullable String resolveValue(@Nullable String s)
	{
		return getOptionalValue(s).orElse(null);
	}

	@Override
	public @Nullable String resolveValue(@Nullable String s, @Nullable String s1)
	{
		return getOptionalValue(s).orElse(s1);
	}
}
```
